### DISCUSSION TITLE
Should TriliumNext explore alternative editors/depart from CKEditor?

### DISCUSSION THREAD

**Discussion by rom1dep** [TriliumNext/Trilium#5907] [Polls]
CKEditor has been very central to Trilium (and in fairness, possibly to its success), but, from a look at the issues tracker, it is also a recurrent point of friction and frustration (I listed [my own](https://github.com/zadam/trilium/issues/1850#issuecomment-1890783396)).

I find it interesting to survey whether departing from CKEditor should be a goal for TrilliumNext or not, especially in light of other opensource note-taking apps often rolling-out their own (some being even great!).

Some examples:
- [siyuan](https://b3log.org/siyuan/en/)
- [logseq](https://github.com/logseq/logseq)
- [joplin](https://github.com/laurent22/joplin/)
- [lexical](https://lexical.dev/) (used by [standardnotes](https://standardnotes.com/))
- [quill](https://quilljs.com/)
- [prosemirror](https://prosemirror.net)
- [tiptap](https://tiptap.dev/)

Some discussion:
https://news.ycombinator.com/item?id=40089460
https://discuss.prosemirror.net/t/differences-between-prosemirror-and-lexical/4557

personal take: siyuan has some great editing capabilities/UX, lexical looks **very** promising

--- 

list (to be continued!) of known issues/limitations with CKEditor that a replacement candidate would rather not exhibit:

- impossible to collapse bullets: https://github.com/zadam/trilium/issues/1850
- caret navigation and selection is broken when references are inserted at the end of the line or immediately after the current caret position
- no stable block reference (impossible to embed blocks/paragraphs across notes): https://github.com/zadam/trilium/issues/4533
- no capability for multi-columns edition: https://github.com/zadam/trilium/discussions/4654

list (to be continued!) of desirable features to have:

- rich text formatting
- content embedding (image, sound, video)
- other notes transclusion
- table of content

edit: added prosemirror and quill

**Comment by eliandoran** [comment]
@rom1dep , I believe at least for now we shouldn't _replace_ it, but keep it alongside.

My dream is to have support for multiple editors. Just like you can choose to have a code note, a Mermaid note, you can have a CKEditor note as well as any other editor you can dream of that we can integrate. This would require a modularization of the application, more explicitly extracting the CKEditor into some kind of plugin and do the same for the new editor. Then we can have Notion-like spreadsheets support and whatever we can dream of.

We also have to consider that existing users' notes are based on CKEditor, which has its own layout in HTML. I think these can't be seamlessly ported to something else without causing potential issues.

The examples you have given seem quite nice. Can you have a look to see which one would be the best candidate? Bonus points for fully open-source ones. I'm a bit tired of seeing a lot of features in CKEditor that are hidden behind a paywall (premium features).

**Reply by rom1dep** [reply]
Agreed with everything you say! The co-existence of multiple editors would also require some careful thoughts in order not to ruin the user experience (after all, the editor should remain mostly an implementation detail).

As I see it, whichever editor comes up to replace CKEditor needs to be the default for new notes and existing ones (even if that means converting them as a one-off migration ; I wonder if going over markdown is feasible/desirable there). Beyond that, specific notes could warrant their specific editor (including CKEditor as an option), but we should manage expectations regarding converting from one to the other.

> Can you have a look to see which one would be the best candidate?

from a quick comparison, [lexical](https://lexical.dev/) seems like a good candidate. It's project page is here: https://github.com/facebook/lexical

**Reply by eliandoran** [reply]
@rom1dep , honestly regarding using a Markdown editor is not the best option in my opinion (hot take?). I feel that the WYSIWYG experience that CKEditor (and potentially other editors offer) is net superior for getting work done. For example I use a lot of tables including images at work and that would be horrendous under Markdown unless you use some kind of hacks such as blocks

I wouldn't want us to degrade the experience of Trilium Notes by using an editor that offers less than what CKEditor already offers.

At a first glance from the [Lexical playground](https://playground.lexical.dev/), it does seem quite interesting. Do you think you could do some kind of comparison of CKEditor vs Lexical? What benefits would we have from switching to it?

**Reply by rom1dep** [reply]
> honestly regarding using a Markdown editor is not the best option in my opinion (hot take?).

not a hot take, I only brought up Markdown as an exchange/intermediate format to explore for conversion. I do think it's important not to force users into markdown. 

> Do you think you could do some kind of comparison of CKEditor vs Lexical? What benefits would we have from switching to it?

I think an interesting thing to do would be to consolidate a "list of problems with CKEditor" from scanning through open issues, and then to see if/which alternatives solve those. I will edit my original post with a short list from the top of my head, and welcome contributions

**Reply by zerebos** [reply]
I like the idea of supporting multiple editors! It would also allow us to support markdown editors for those that prefer it without us having to compromise on WYSIWYG for the vast majority of users!

**Reply by bwat47** [reply]
WYSIWYG is one of the primary reasons I use trillium over the sea of markdown-based notes programs out there. I strongly dislike markdown.

For example, I work in technical support and I use trillium for my personal knowledgebase of technical support notes. This includes many canned replies that I often copy into emails. If I copy formatted text and paste it in an email, I need the formatted text. 

I've tried so many notes programs where it will display the formatted text in the notes program, but when pasting into an email, I get the plain markdown. I then promptly uninstall that program and go back to trillum :D

**Reply by Byteingpython** [reply]
Having an editor with great markdown support doesn't mean not being WYSIWYG. Even the current editor has some level of markdown conversion. And as you already said it is a WYSIWYG editor

**Reply by BeatLink** [reply]
I like this idea a lot. Modular notes are the way to go.

**Comment by meichthys** [comment]
I can also say that the Siyuan editing experience is very nice.

**Reply by cjrutherford** [reply]
Siyuan looks like if Obsidian and Notion had offspring. that's not a bad thing.

**Comment by shutaozhenzhen** [comment]
Ckeditor is clearly the source of many of trilium's problems, but it's clear that replacing Ckeditor isn't the most important thing right now, and it would be a major change that should be dealt with after it's stabilized (and worth releasing a new version).
Moreover, many online editors can available as plugins to trilium, making the problem less pressing.

**Reply by cjrutherford** [reply]
Agreed with this, it's a consideration between improving the experience for the user AND the developer. Have the issues been transferred? I'd like to see what the problems are and try to understand why they're there.

**Reply by meichthys** [reply]
https://github.com/TriliumNext/Notes/issues/79

**Comment by flysaber** [comment]
On the contrary, I think the editor experience of Siyuan is very poor and not smooth at all.

**Reply by eliandoran** [reply]
@flysaber , sounds interesting. Could you share some examples?

**Comment by Dialga** [comment]
Add https://github.com/AppFlowy-IO/AppFlowy to your list, they have their own editor.
Edit: see also https://github.com/streetwriters/notesnook.

**Reply by rom1dep** [reply]
It doesn't look like they package their editor as a separate library, or am I missing something? If not, it might become extremely difficult to vendor in a project like this.

**Reply by b-jordanov** [reply]
I have tried Appflowy for a good amount and moved to Trillium. Their editor is intuitive. I love it that I don't have to use the mouse just press / and select what type of block I need with the keyboard .

I like that we can keep the notes in .md and organize them and I also notice the CK Editor limitations. I would love to overcome them. I also tried siyuan and I liked it a lot, except it comes with it's own boundaries in how the UI looks and acts.

Even tough they went FOSS a while ago I'd still prefer to stick with Trillium at least for now but I'm glad to see a discussion about a better editor because it's definitely a must.

**Comment by Alumniminium** [comment]
I wish trilium had Dillinger as edtior. https://dillinger.io/ 

I just want to write markdown and have proper markdown rendering

**Comment by CrO2Cl2** [comment]
I do also think that Trilium should have a LaTeX-first editor, for the ones of use that nearly only write in LaTeX. One of the renderers I can think of that work really well is [Mathpix Markdown](https://github.com/Mathpix/mathpix-markdown-it). shouldnt be too hard to implement, and is documented very well. SMILES support is also a plus

**Comment by arcqus** [comment]
Hello
Joplin's editor is just a shame (and the reason I'm here). It is full of bugs (add or remove lines randomly, bullets badly managed...) and very poor from a end-user standpoint.
As far as I understand, it is related to the use of markdown. 
Personally, I'm looking for a MS OneNote-like solution, and trilium was the closest I got.

**Comment by Byteingpython** [comment]
I think a new default editor should at least have good support for people prefering their keyboard over their mouse. For me that would be full markdown compatibility and a vim mode. It can be WYSIWYG but it must support automatically converting every part of the markdown syntax to its own format.

**Reply by b-jordanov** [reply]
I completely agree, touching the mouse loses me so much time but the most important part of this reply is "it must support automatically converting every part of the markdown syntax to its own format"

**Reply by XingSc29** [reply]
Agree too

**Comment by meichthys** [comment]
Issue created for this: https://github.com/TriliumNext/Notes/issues/79

**Comment by meichthys** [comment]
Some of the pending feature requests are at least somewhat limited by our current editor (CKEditor5). For the most part, CKEditor does support the functionality to implement some of these feature requests, but the needed CKEditor functionalities are [premium features](https://ckeditor.com/ckeditor-5/features/) that are behind a paywall. 

However! CKEditor states the following on their open source page:

> ### How we support Open Source
> 
> We grant a free CKEditor 5 license to Open Source projects released under a license incompatible with GPL2+.
> 
>     - Terms & Conditions:
>     - Your project needs to be released under an [OSI-approved license](https://opensource.org/licenses/category).
>     - Your source code needs to be publicly available.
>     - Your software must not enter into direct competition with our products.
>     - You must contact us to obtain a license.
> 

It would take a bit more research to see if we qualify and what features we would get in this OpenSource License, but it sure is worth investigating.

Some of the feature requests that could be solved by a 'fuller' CKEditor are:

- https://github.com/TriliumNext/Notes/issues/97 
  - https://ckeditor.com/docs/ckeditor5/latest/getting-started/setup/ui-language.html#righttoleft-rtl-languages-support
- Document outline / TOC (We have this already, but perhaps this would be more 'native')
  - https://ckeditor.com/docs/ckeditor5/latest/features/document-outline.html
- https://github.com/TriliumNext/Notes/issues/73
  - https://ckeditor.com/docs/ckeditor5/latest/features/math-equations.html
- Slash Commands
  - https://ckeditor.com/docs/ckeditor5/latest/features/slash-commands.html
- Grammar Check
  - https://ckeditor.com/docs/ckeditor5/latest/features/spelling-and-grammar-checking.html
- AI features
  - https://ckeditor.com/docs/ckeditor5/latest/features/ai-assistant/ai-assistant-overview.html
- Collaboration
  - https://ckeditor.com/docs/ckeditor5/latest/features/collaboration/real-time-collaboration/real-time-collaboration.html?docId=e4b3de894eeff72d0669e1d176c84766c

**Reply by CrO2Cl2** [reply]
it sounds like a really cool idea. It would also finally mean that we support full LaTeX, collaboration ( when we add multi user ), and better exports (as reqested not too long ago),

**Reply by eliandoran** [reply]
@meichthys , I think the license itself is only to gain access to the base functionality, not the premium ones. For premium features it says:

> This is a premium feature, and you need a license for it on top of your CKEditor 5 commercial license. [Contact us](https://ckeditor.com/contact/?sales=true#contact-form) to receive an offer tailored to your needs.

But indeed it might be worth contacting them to see if they have an offer for premium features for open-source and how it would work. I'll give it a try.

**Reply by meichthys** [reply]
I remember I asked Zadam about this as well in the past and I don't think he ever got a license from CK editor so this might be something different (or at least i hope so!)

**Reply by meichthys** [reply]
Here is a better description of their licensing on their forum: https://support.ckeditor.com/hc/en-us/articles/115002452529-Under-which-licenses-does-CKEditor-come

**Reply by CobriMediaJulien** [reply]
Did sb. allrdy contact them?

**Reply by meichthys** [reply]
I went ahead ans submitted an application here: https://info.ckeditor.com/l/1018062/2024-09-12/3hvct
I'll follow up when I hear back. Unfortunately we'll all have to take pay cuts since we won't qualify if we make more than $2m/yr 🤣

**Comment by kryoqa** [comment]
I am not a developer but can any of the following be used for trilium ?

[https://github.com/toeverything/blocksuite](Blocksuite - Affine)

[https://github.com/treehousedev/treehouse](Treehouse - Front end similar to Tana)

**Comment by CyberFoxar** [comment]
I'm throwing another text editor for your consideration [tiptap](https://tiptap.dev/product/editor) !

Things that it has going for it compared to CKEditor:
- JSON-based schema (like Lexical)
- Easier to extend
- More mature than Lexical (but slower, less shiny)
- More upfront work (No ready-to-use toolbars, lots of community-proved ones)
- Built ontop of ProseMirror (might consider using that instead)
